Document Outlining Sequence of a Single Library Migration (#235)
ROCm/rocm-libraries: docs/migration-process.md

# Migration from Single Repo to Monorepo  
  
## Introduction  
This document outlines the process for migrating from a single library repository to this monorepo. It covers the necessary steps to ensure a smooth transition, including pre-conditions, conflict resolution, and changes to repository management.  
  
## Pre-conditions  
To ensure consistency and maintainability during the migration, the following pre-conditions must be satisfied:  

1. **Identify Next Repo to Migrate:**

>    - Please refer to the main [README.md](README.md) on the order of repositories being migrated.

Reviewer: Looks like this link is broken, probably need to add a `..`

Author: Fixed

# Migration from Single Repo to Monorepo  
  
## Introduction  
This document outlines the process for migrating from a single library repository to this monorepo. It covers the necessary steps to ensure a smooth transition, including pre-conditions, conflict resolution, and changes to repository management.  
  
## Pre-conditions  
To ensure consistency and maintainability during the migration, the following pre-conditions must be satisfied:  

1. **Identify Next Repo to Migrate:**
   - Please refer to the main [README.md](/README.md) on the order of repositories being migrated.
   - This is usually discussed in advance in meetings with the technical leads of that project.

2. **Identify Branches and Pull Requests:**  
   - Determine branches and active pull requests that will be affected by the migration.
   - Typically, this is limited to the pull requests targeting `develop` and `release-staging` branches.
   - Any point-fixes for previous releases will not be migrated over.

3. **Pause Merges:**
   - There are GitHub Actions that automatically synchronize changes from the individual repos to the monorepo.

>    - These automated actions need to be paused by selecting to disable the workflow on the GitHub UI.

Reviewer 2: ```suggestion
   - These automated actions need to be paused by disabling the workflow on the GitHub UI.
```

   - develop branch workflow: https://github.com/ROCm/rocm-libraries/actions/workflows/update-subtrees.yml 
   - release-staging branch workflow: https://github.com/ROCm/rocm-libraries/actions/workflows/update-release-staging-subtree.yml

>    - Announce the pause to key stakeholders, and ask to propagate the news.

Reviewer 2: ```suggestion
   - Announce the pause to key stakeholders and ask them to propagate the news.
```

# Delete the temporary branches created in this sequence.
```

### Step 4: CI/CD Triggers
  
1. **CI/CD Trigger Points:**
   - Modify the existing CI/CD systems to be triggered off changes to this project in the monorepo.

### Step 5: Repository Adjustments

1. **Default Branch Deprecation:**  
   - Change the default branch of the original repository with a clear deprecation notice.
   - Example: https://github.com/ROCm/rocPRIM/tree/develop_deprecated
  
2. **Disable Dependabot Updates:**  
   - Cease automatic dependency updates in the old repository to streamline the focus on the monorepo.
   - Clear the contents in this file on the original repo: https://github.com/ROCm/rocPRIM/blob/develop_deprecated/.github/dependabot.yml
   - In the original repo settings, go to Security -> Advanced Security and disable all the Dependabot settings.

3. **Protection Rules:**
   - Use branch protection to make the new default branch with the deprecation notice read-only.
   - Create a ruleset for the `develop` branch to also be restrictive, but allow the assistant-librarian bot exceptions to push patches to the original repository.

### Step 6: Source of Truth Declaration

> 1. **Update repo config json:**

Reviewer 2: Is it repo config, repo-config, or repos-config (I see all three)?

Author: It's repos-config.json, I'll make it consistent and link the file.

>    - Update the true/false values in the repo-config json file that automated workflows use to determine which way source gets synchornized.

Reviewer 2: ```suggestion
   - Update the true/false values in the repo-config json file that automated workflows use to determine which way the source gets synchronized.
```

Author: Corrected
